libp2p/NodeTable.cpp

```diff
@@ -557,19 +562,31 @@ void NodeTable::doCheckEvictions()
         
         bool evictionsRemain = false;
         list<shared_ptr<NodeEntry>> drop;
+        list<shared_ptr<NodeEntry>> active;
+
         {
             Guard le(x_evictions);
             Guard ln(x_nodes);
             for (auto& e: m_evictions)
                 if (chrono::steady_clock::now() - e.second.evictedTimePoint > c_reqTimeout)
-                    if (m_nodes.count(e.second.newNodeID))
-                        drop.push_back(m_nodes[e.second.newNodeID]);
+                    if (m_nodes.count(e.first))
+                    {
+                        drop.push_back(m_nodes[e.first]);
+                        if (m_nodes.count(e.second.newNodeID))
+                            active.push_back(m_nodes[e.second.newNodeID]);
+                    }
             evictionsRemain = (m_evictions.size() - drop.size() > 0);
+
+            drop.unique();
+            for (auto n: drop)
+            {
+                m_evictions.erase(n->id);
+                dropNode(n);
```

Comment:
I think this will deadlock. `x_nodes` is already locked, then your version of `dropNode` will try to do it once again.
No need to call `dropNode` here inside `x_evictions` & `x_nodes` locks, I'd suggest to move this loop back, but do `m_evictions.erase(n->id);` separately. It's possible to iterate `m_evictions` erasing elements on the way.